### PR TITLE
Fix context missing for `yarn build` errors

### PR DESCRIPTION
**Pull request by mitchhentgesspotify**
Make build error cause more clear by prepending with module name

Simplify the diagnosis of build failures by showing the module name
encountering the issue right before the error itself is printed.

Also apply the fix to warnings.

Finally, while in the area, remove a `react-dev-utils` workaround which should no longer be needed.

-----

Before:
```
Error: Failed to compile.
Attempted import error: 'smile' is not exported from '@backstage/plugin-catalog-common/alpha' (imported as 'smile').
```
After:
```
Error: Failed to compile.
./src/App.tsx:
  Attempted import error: 'smile' is not exported from '@backstage/plugin-catalog-common/alpha' (imported as 'smile').
```

-----

See [the webpack docs here](https://webpack.js.org/api/stats/#errors-and-warnings) for the structure of error and warnings objects. 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: packages/cli/src/lib/bundler/bundle.ts [hunk 1]
```diff
@@ -32,6 +32,10 @@ import chalk from 'chalk';
 const WARN_AFTER_BUNDLE_GZIP_SIZE = 512 * 1024;
 const WARN_AFTER_CHUNK_GZIP_SIZE = 1024 * 1024;
 
+function applyContextToError(error: string, moduleName: string): string {
```
Comment (Rugvip) [inline packages/cli/src/lib/bundler/bundle.ts:35]: I think you flipped the params around here 😅

**Review by Rugvip — CHANGES_REQUESTED** [review]
Nice! Yeah this'll definitely help identify errors easier

File: packages/cli/src/lib/bundler/bundle.ts (hunk 1 above)
Comment (mitchhentgesspotify) [reply packages/cli/src/lib/bundler/bundle.ts:35]: Whoops, a little end-of-day distracted refactoring. Thanks for the catch :sweat_smile:

**Comment by mitchhentgesspotify** [comment]
New output:

```
$ yarn build
Loaded config from app-config.yaml

Treating warnings as errors because process.env.CI = true.


Error: Failed to compile './src/App.tsx':
  Attempted import error: 'nonexistentProperty' is not exported from '@backstage/plugin-catalog-common/alpha' (imported as 'nonexistentProperty').
```

File: packages/cli/src/lib/bundler/bundle.ts [hunk 2]
```diff
@@ -114,37 +115,41 @@ async function build(config: webpack.Configuration, isCi: boolean) {
...
+      errors[0],
+      serializedStats.errors?.[0].moduleName || '',
+    );
+    throw new Error(errorWithContext);
   }
   if (isCi && warnings.length) {
+    const warningsWithContext = warnings.map((warning, i) => {
+      return applyContextToError(
+        warning,
+        serializedStats.warnings?.[i].moduleName || '',
```
Comment (benjdlambert) [inline packages/cli/src/lib/bundler/bundle.ts:144]: Maybe we should do something like:

```suggestion
        serializedStats.warnings?.[i]?.moduleName ??  '',
```

File: packages/cli/src/lib/bundler/bundle.ts [hunk 3]
```diff
@@ -114,37 +115,41 @@ async function build(config: webpack.Configuration, isCi: boolean) {
...
+    warnings: serializedStats.warnings,
   });
 
   if (errors.length) {
     // Only keep the first error. Others are often indicative
     // of the same problem, but confuse the reader with noise.
-    throw new Error(errors[0]);
+    const errorWithContext = applyContextToError(
+      errors[0],
+      serializedStats.errors?.[0].moduleName || '',
```
Comment (benjdlambert) [inline packages/cli/src/lib/bundler/bundle.ts:136]: Maybe same as below here?

**Review by benjdlambert — COMMENTED** [review]
Just a quick comment, indexing arrays make me nervous when they're not the same array that you're interating over in some cases? 👍

**Review by benjdlambert — APPROVED** [review]

**Review by Rugvip — APPROVED** [review]

**Comment by mitchhentgesspotify** [comment]
> Just a quick comment, indexing arrays make me nervous when they're not the same array that you're interating over in some cases? +1

That's fair, I reached for a "array zip"-style function for better ergonomics, but opted to keep it simple (and less "obviously correct", as you mentioned) since I couldn't find such a tool in the standard lib.

> serializedStats.warnings?.[i]?.moduleName ??  ''

I'm surprised that `tsc` didn't mention that indexing `warnings` could return `undefined` :thinking: 
Thanks for following up on the PR and landing it 👍

**Pull request merged**